modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EChain.java

```diff
@@ -215,15 +215,30 @@ private void analyzeCompound(final CompilerSettings settings, final Definition d
         there = AnalyzerCaster.getLegalCast(definition, location, last.after, promote, false);
         back = AnalyzerCaster.getLegalCast(definition, location, promote, last.after, true);
 
+        if (last instanceof ADefLink) {
+            final ADefLink lastDef = (ADefLink) last;
+            // Unfortunately, we don't know the real type because we load from DEF and store to DEF!
+            lastDef.storeValueType = last.after;
+        }
+
         statement = true;
         actual = read ? last.after : definition.voidType;
     }
 
     private void analyzeWrite(final CompilerSettings settings, final Definition definition, final Variables variables) {
         final ALink last = links.get(links.size() - 1);
 
-        expression.expected = last.after;
-        expression.analyze(settings, definition, variables);
+        // If the store node is a DEF node, we remove the cast to DEF from the expression
+        // and promote the real type to it:
+        if (last instanceof ADefLink) {
+            final ADefLink lastDef = (ADefLink) last;
+            expression.analyze(settings, definition, variables);
+            lastDef.storeValueType = expression.expected = expression.actual;
```

Comment:
I tried it out, works without any problems! I never change the actual type of the expressions. And the value is duped much later. And the bytecode shows that ir correctly inserts the cast afterwards (on the duped value):

`int x; def[] y = new def[1]; x = y[0] = 1; return x;`

Yields the following bytecode, with cast after the array store:

```
  public execute(Ljava/util/Map;Lorg/apache/lucene/search/Scorer;Lorg/elasticsearch/search/lookup/LeafDocLookup;Ljava/lang/Object;)Ljava/lang/Object;
   L0
    LINENUMBER 1 L0
    ICONST_0
    ISTORE 5
   L1
    LINENUMBER 1 L1
    ICONST_1
    ANEWARRAY java/lang/Object
    ASTORE 6
   L2
    LINENUMBER 1 L2
    ALOAD 6
    ICONST_0
    ICONST_1
    INVOKESTATIC java/lang/Integer.valueOf (I)Ljava/lang/Integer;
    DUP_X2
    AASTORE
    INVOKESTATIC org/elasticsearch/painless/Def.DefToint (Ljava/lang/Object;)I
    ISTORE 5
   L3
    LINENUMBER 1 L3
    ILOAD 5
    INVOKESTATIC java/lang/Integer.valueOf (I)Ljava/lang/Integer;
    ARETURN
    MAXSTACK = 4
    MAXLOCALS = 7
}
```